Project/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs

```diff
@@ -184,64 +266,87 @@ void FixedUpdate()
+                $" maximumWalkingSpeed: {m_maxWalkingSpeed}"
+            );
         }
 
-        if (rewardUseTimePenalty)
+        // b. Rotation alignment with target direction.
+        //This reward will approach 1 if it faces the target direction perfectly and approach zero as it deviates
+        var lookAtTargetReward = (Vector3.Dot(cubeForward, body.forward) + 1) * .5F;
+
+        //Check for NaNs
+        if (float.IsNaN(lookAtTargetReward))
         {
-            RewardFunctionTimePenalty();
+            throw new ArgumentException(
+                "NaN in lookAtTargetReward.\n" +
+                $" cubeForward: {cubeForward}\n" +
+                $" body.forward: {body.forward}"
+            );
         }
+
+        AddReward(matchSpeedReward * lookAtTargetReward);
     }
 
-    /// <summary>
-    /// Reward moving towards target & Penalize moving away from target.
-    /// </summary>
-    void RewardFunctionMovingTowards()
+    //Update OrientationCube and DirectionIndicator
+    void UpdateOrientationObjects()
     {
-        var movingTowardsDot = Vector3.Dot(orientationCube.transform.forward,
-            Vector3.ClampMagnitude(m_JdController.bodyPartsDict[body].rb.velocity, maximumWalkingSpeed));
-        if (float.IsNaN(movingTowardsDot))
+        m_OrientationCube.UpdateOrientation(body, m_Target);
+        if (m_DirectionIndicator)
         {
-            throw new ArgumentException(
-                "NaN in movingTowardsDot.\n" +
-                $" orientationCube.transform.forward: {orientationCube.transform.forward}\n" +
-                $" body.velocity: {m_JdController.bodyPartsDict[body].rb.velocity}\n" +
-                $" maximumWalkingSpeed: {maximumWalkingSpeed}"
-            );
+            m_DirectionIndicator.MatchOrientation(m_OrientationCube.transform);
         }
-        AddReward(0.03f * movingTowardsDot);
     }
 
-    /// <summary>
-    /// Reward facing target & Penalize facing away from target
-    /// </summary>
-    void RewardFunctionFacingTarget()
+    //Returns the average velocity of all of the body parts
+    //Using the velocity of the hips only has shown to result in more erratic movement from the limbs, so...
+    //...using the average helps prevent this erratic movement
+    Vector3 GetAvgVelocity()
     {
-        var facingReward = Vector3.Dot(orientationCube.transform.forward, body.forward);
-        if (float.IsNaN(facingReward))
+        Vector3 velSum = Vector3.zero;
+        Vector3 avgVel = Vector3.zero;
+
+        //ALL RBS
+        int numOfRB = 0;
+        foreach (var item in m_JdController.bodyPartsList)
         {
-            throw new ArgumentException(
-                "NaN in movingTowardsDot.\n" +
-                $" orientationCube.transform.forward: {orientationCube.transform.forward}\n" +
-                $" body.forward: {body.forward}"
-            );
+            numOfRB++;
+            velSum += item.rb.velocity;
```

Comment:
no changes. i considered doing this but was hoping it would help smooth the jittery legs. if we weight it by mass i assumed it wouldn't work as well but i may be wrong.

Reply:
Seems counter intuitive to me that the fingers' speed would contribute as much to overall speed as the torso's. This weights the number of rigidbodies more than their mass.

Reply:
i agree but i want to normalize/smooth movement as much as possible and my approach attempts to convey that the velocity of the hands should be as important as velocity of the larger body parts. We could make a test branch to test the difference between the two.